Microsoft.AzureStackHCI API: 2025-12-01-preview

shreyanshd opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.
- [ ] A [release plan](https://aka.ms/azsdkdocs/release-plans) has been created. If not, please create one as it will help guide you through the REST API and SDK creation process. 

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.
- For guidance on SDK breaking change review, refer to https://aka.ms/ci-fix.

Commit:
Copy from 2025-11-01-preview

Commit:
Update readme

Commit:
replace API version

Commit:
replace API version for examples

Commit:
add new models, interfaces, examples

Commit:
Fix examples

Commit:
fix swagger and examples

Commit:
fixes

Commit:
fixes

Commit:
fixes typespec vlidation

abaranwal-ms committed:
Merge branch 'main' into users/shreyanshd/hci-api-2025-12-01-preview

abaranwal-ms committed:
Users/abaranwal/verify changes hci nov14 (#38809)

* sdn, cvm properties

* lastSyncTimestamp and Example changes

abaranwal-ms committed:
re-order models for readability

sandipsh commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/StackHCI/edgeMachineJob.tsp:
```diff
@@ -0,0 +1,761 @@
+
+  @doc("The timestamp when log collection ended, in ISO 8601 format.")
+  @visibility(Lifecycle.Read)
+  endTime?: string;
```
why not utc

sandipsh commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/StackHCI/edgeMachineJob.tsp:
```diff
@@ -0,0 +1,761 @@
+  secretLocation?: string;
+
+  @doc("Location of the secret used for authentication.")
+  sshPubKey?: string[];
```
>sshPubKey [](http://example.com/codeflow?start=2&length=9)

needs to be plural

mentat9 replied:
At least fix the description. Apparently it's a collection of keys?

sandipsh commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/StackHCI/edgeMachineJob.tsp:
```diff
@@ -0,0 +1,761 @@
+  connectionUri?: string;
+
+  @doc("Port of the web proxy.")
+  port?: string;
```
>port?: string; [](http://example.com/codeflow?start=2&length=14)

isn't there any pattern needed for this?

shreyanshd replied:
added pattern

sandipsh commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/StackHCI/preview/2025-12-01-preview/edgeDevices.json:
```diff
@@ -0,0 +1,1399 @@
+        "lastSyncTimestamp": {
+          "type": "string",
+          "format": "date-time",
+          "description": "Most recent edge device sync timestamp.",
```
>Most recent edge device sync timestamp. [](http://example.com/codeflow?start=26&length=39)

please mention if it is in Utc.

sandipsh commented:
>         "Unknown"

this should be 1st in the list.

---
Refers to: specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/StackHCI/preview/2025-12-01-preview/hci.json:4038 in 82f8fc6. [](commit_id = 82f8fc6b58c97f7ee2375340d3006b779d7ee702, deletion_comment = False)

abaranwal-ms committed:
Re-order enum as suggested, fix RDMA description, add UTC description over timestamp field

abaranwal-ms committed:
Merge branch 'main' into users/shreyanshd/hci-api-2025-12-01-preview

Commit:
Add minItems

Commit:
Merge branch 'users/shreyanshd/hci-api-2025-12-01-preview' of https://github.com/Azure/azure-rest-api-specs into users/shreyanshd/hci-api-2025-12-01-preview

Commit:
improve validations

Commit:
Add edge machine patch

Commit:
Remove labels

shreyanshd commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/StackHCI/edgeMachineJob.tsp:
```diff
@@ -0,0 +1,761 @@
+  secretLocation?: string;
+
+  @doc("Location of the secret used for authentication.")
+  sshPubKey?: string[];
```
This is a SSH public key and not a secret

shreyanshd commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/StackHCI/edgeMachine.tsp:
```diff
@@ -0,0 +1,899 @@
+model ReportedProperties {
+  @doc("Last time data reported.")
+  @visibility(Lifecycle.Read)
+  lastUpdated?: utcDateTime;
```
As part of approved PR: https://github.com/Azure/azure-rest-api-specs-pr/pull/23165/ in private repo, the device agent SDK has taken a dependency on this property and changing it would result in breaking change for the product in public preview. Hence, keeping it consistent with the approved private preview API version

abaranwal-ms committed:
Merge branch 'main' into users/shreyanshd/hci-api-2025-12-01-preview

ravimeda commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/StackHCI/edgeMachine.tsp:
```diff
@@ -0,0 +1,813 @@
+model StorageProfile {
+  @doc("Number of storage disks in the device with $CanPool as true.")
+  @visibility(Lifecycle.Read)
+  poolableDisksCount?: int64;
```
>poolableDisksCount [](http://example.com/codeflow?start=2&length=18)

Assuming the business logic will ensure the values is not negative

shreyanshd replied:
Business logic handles the validation

ravimeda commented on specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/StackHCI/edgeMachineJob.tsp:
```diff
@@ -0,0 +1,741 @@
+  RemoteSupport: "RemoteSupport",
+
+  @doc("Job to provision operating system in the device.")
+  ProvisionOs: "ProvisionOs",
```
>Os [](http://example.com/codeflow?start=25&length=2)

nit - Only when the acronym is three characters or more, PascalCase is used. Here I would suggest `ProvisionOS`

Commit:
Address review comments

Commit:
Merge branch 'users/shreyanshd/hci-api-2025-12-01-preview' of https://github.com/Azure/azure-rest-api-specs into users/shreyanshd/hci-api-2025-12-01-preview

abaranwal-ms committed:
Merge branch 'main' into users/shreyanshd/hci-api-2025-12-01-preview

shreyanshd commented:
> @shreyanshd - I'm not sure I understand exactly what's happening with this PR. Is this just to promote an existing API version (already signed off by ARM) from the private repo to public? If so, please indicate that when you submit the PR (along with a link to the previous PR). This saves reviewers a lot of time: we don't want to re-review the same APIs that have already shipped publicly. If this is adding new APIs, then you should be free to change things like property names. Can you clarify what this PR is for?

@mentat9 - We have an approved PR: https://github.com/Azure/azure-rest-api-specs-pr/pull/23165 for API version `2024-11-01-preview`. This is a private preview API, and our product was in private preview for past one year. Now we intend to promote the resource types to public preview with this API version `2025-12-01-preview`. So, this PR includes:

1. `EdgeMachine` and `EdgeMachineJobs` resource types which were part of approved private preview PR: https://github.com/Azure/azure-rest-api-specs-pr/pull/23165. Since these models were already in public preview, Device SDKs have already taken a dependency on them and changing them would mean a breaking change for all private preview users who would continue the same product in public preview. Having said that, all other review comments except property renaming has been addressed.
2. `ValidateOwnershipVouchers` resource - this is a new resource that we have added in this API version.

abaranwal-ms committed:
Merge branch 'main' into users/shreyanshd/hci-api-2025-12-01-preview

abaranwal-ms committed:
baseline commit, bring equivalent to previous preview: 2025-12-01-preview

abaranwal-ms committed:
copying approved changes from private branch

abaranwal-ms committed:
Bring changes for review

shreyanshd commented:
> > > > @shreyanshd - I'm not sure I understand exactly what's happening with this PR. Is this just to promote an existing API version (already signed off by ARM) from the private repo to public? If so, please indicate that when you submit the PR (along with a link to the previous PR). This saves reviewers a lot of time: we don't want to re-review the same APIs that have already shipped publicly. If this is adding new APIs, then you should be free to change things like property names. Can you clarify what this PR is for?
> > > 
> > > 
> > > @mentat9 - We have an approved PR: [Azure/azure-rest-api-specs-pr#23165](https://github.com/Azure/azure-rest-api-specs-pr/pull/23165) for API version `2025-08-01-preview`. This is a private preview API, and our product was in private preview for past one year. Now we intend to promote the resource types to public preview with this API version `2025-12-01-preview`. So, this PR includes:
> > > 
> > > 1. `EdgeMachine` and `EdgeMachineJobs` resource types which were part of approved private preview PR: [AzureStackHCI 2025-08-01-preview azure-rest-api-specs-pr#23165](https://github.com/Azure/azure-rest-api-specs-pr/pull/23165). Since these models were already in public preview, Device SDKs have already taken a dependency on them and changing them would mean a breaking change for all private preview users who would continue the same product in public preview. Having said that, all other review comments except property renaming has been addressed.
> > > 2. `ValidateOwnershipVouchers` resource - this is a new resource that we have added in this API version.
> > 
> > 
> > Thank-you, this helps a bit, but the refactoring and moving of files makes it really hard to track what's changing. For example, currently in the -pr repo, I expected to see spec files edgeMachine.tsp and edgeMachineJobs.tsp so I could compare them to files in this PR, but they aren't there. The PR you linked shows them being added, so they have been deleted afterward. Comparing those two files in the earlier PR to this PR shows differences. Can you explain what's going on with those two resource types? Where did you get those files when you added them to this PR and why are they no longer part of the private spec?
> > This PR is a refactoring taking place at the same time you are moving from private to public repo and also adding APIs in a new API version. Reviewing this properly is essentially impossible. Can you do the promotion from private to public repo, refactoring, and new API version as separate PRs? Basically, we should be able to see the spec copied unchanged from private to public repo. Then a refactoring that doesn't change API versions, then a new API version to add new APIs (or in some other order if that works better).
> 
> @mentat9 - We understand the confusion this might have caused to the reviewers. To simplify the review process for this PR, we have updated the PR with following 3 commits:
> 
> Commit	Details
> [766529](https://github.com/Azure/azure-rest-api-specs/pull/38737/commits/76652974e2d76c52a8fbbb31444b22b94c434e68)	Baseline commit from `2025-11-01-preview`
> [a5049d](https://github.com/Azure/azure-rest-api-specs/pull/38737/commits/a5049d71e275f0469dda748639c1f2ec6fb50fe8)	Bring changes from private branch (approved PR: [Azure/azure-rest-api-specs-pr#23165](https://github.com/Azure/azure-rest-api-specs-pr/pull/23165))
> [7555fb](https://github.com/Azure/azure-rest-api-specs/pull/38737/commits/7555fbe0499cb7471988f21cb7470e3feda48d37)	This commit has all the additional changes introduced in `2025-12-01-preview`
> The last commit essentially has all the changes which needs to be reviewed, hope this helps!

<b>Note to reviewer</b>: The Cluster, EdgeDevice, EdgeMachine, EdgeMachineJobs models are also used by the device SDK and tooling. Many attributes in these models already have established dependencies. Renaming such attributes would be difficult, as we are not in a position to upgrade the device-side tooling to newer contracts. Therefore, we aim to maintain backward compatibility.

abaranwal-ms committed:
Merge branch 'main' into users/shreyanshd/hci-api-2025-12-01-preview

abaranwal-ms committed:
remove added for preview version for arm versioning policy

abaranwal-ms committed:
Merge branch 'users/shreyanshd/hci-api-2025-12-01-preview' of https://github.com/Azure/azure-rest-api-specs into users/shreyanshd/hci-api-2025-12-01-preview

abaranwal-ms committed:
remove 11-01 version from main.tsp, following versioning policy

abaranwal-ms committed:
add versions as recommended

abaranwal-ms committed:
formatted

abaranwal-ms committed:
Merge branch 'main' into users/shreyanshd/hci-api-2025-12-01-preview

abaranwal-ms committed:
revert to initial state

abaranwal-ms committed:
Merge branch 'users/shreyanshd/hci-api-2025-12-01-preview' of https://github.com/Azure/azure-rest-api-specs into users/shreyanshd/hci-api-2025-12-01-preview
